File: 559.maximum-depth-of-n-ary-tree.java
/*
 * @lc app=leetcode id=559 lang=java
 *
 * [559] Maximum Depth of N-ary Tree
 */

// @lc code=start
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

import java.util.ArrayDeque;
import java.util.Queue;

class Solution {
    public int maxDepth(Node root) {
        if (root == null)
            return 0;
        Queue<Node> queue = new ArrayDeque<>();
        int depth = 0;
        queue.add(root);
        while (!queue.isEmpty()) {
            depth += 1;
            for (int sz = queue.size(); sz != 0; --sz) {
                Node n = queue.remove();
                for (Node child : n.children) {
                    if (child != null)
                        queue.add(child);
                }
            }
        }
        return depth;
    }
}
// @lc code=end

File: 559.maximum-depth-of-n-ary-tree.cpp
/*
 * @lc app=leetcode id=559 lang=cpp
 *
 * [559] Maximum Depth of N-ary Tree
 */

// @lc code=start
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

#include <vector>
#include <algorithm>
#include <queue>

using namespace std;

class Solution {
public:
    int maxDepth(Node* root) {
        if (root == nullptr) return 0;
        queue<Node *> qu;
        qu.push(root);
        int depth = 0;
        while (!qu.empty()) {
            depth += 1;
            for (int sz = qu.size(); sz != 0; --sz) {
                Node *node = qu.front(); qu.pop();
                vector<Node *> children = node->children;
                for_each(children.begin(), children.end(), [&qu](Node *n) {
                    qu.push(n);
                });
            }
        }
        return depth;
    }
};
// @lc code=end
